nx/lib/nx/binary_backend.ex

```diff
@@ -476,8 +476,89 @@ defmodule Nx.BinaryBackend do
   end
 
   @impl true
-  def dot(out, %{type: t1} = left, axes1, %{type: t2} = right, axes2) do
-    bin_zip_reduce(out, left, axes1, right, axes2, 0, fn lhs, rhs, acc ->
+  def dot(out, left, contract_axes1, [], right, contract_axes2, []) do
+    # dot/4 is directed to this specific clause so we can keep a more efficient implementation
+    # for non-batched dot products. See the clause below for batched dot products
+    dot4(out, left, contract_axes1, right, contract_axes2)
+  end
+
+  def dot(
+        out,
+        %{shape: left_shape, type: {_, left_size}, names: left_names} = left,
+        left_contract_axes,
+        left_batch_axes,
+        %{shape: right_shape, type: {_, right_size}, names: right_names} = right,
+        right_contract_axes,
+        right_batch_axes
+      ) do
+    left_binary = to_binary(left)
+    right_binary = to_binary(right)
+
+    left_batch_contract_axes = Nx.Shape.shift_axes(left_contract_axes, -length(left_batch_axes))
+
+    right_batch_contract_axes =
+      Nx.Shape.shift_axes(right_contract_axes, -length(right_batch_axes))
+
+    {left_batch_shape, left_batch_names} =
+      Nx.Shape.contract(left_shape, left_batch_axes, left_names, false)
+
+    {right_batch_shape, right_batch_names} =
+      Nx.Shape.contract(right_shape, right_batch_axes, right_names, false)
+
+    batch_item_length = Nx.size(left_batch_shape)
+
+    batch_count = Nx.Shape.batch_count(left_shape, left_batch_axes)
+
+    {batch_item_output_shape, _output_names} =
+      Nx.Shape.zip_reduce(
+        left_batch_shape,
+        left_batch_contract_axes,
+        left_batch_names,
+        right_batch_shape,
+        right_batch_contract_axes,
+        right_batch_names
+      )
+
+    range = if batch_count == 0, do: [], else: 0..(batch_count - 1)
+
+    left_batch_item_template = %{left | shape: left_batch_shape}
+    right_batch_item_template = %{right | shape: right_batch_shape}
+
+    batch_item_output_template = %{out | shape: batch_item_output_shape}
+
+    bin_result =
+      for index <- range, into: <<>> do
+        offset = index * batch_item_length
+
+        left_offset_bits = offset * left_size
+        right_offset_bits = offset * right_size
+
+        left_batch_item_bits = batch_item_length * left_size
+        right_batch_item_bits = batch_item_length * right_size
+
+        <<_::bitstring-size(left_offset_bits),
+          left_batch_item_binary::bitstring-size(left_batch_item_bits),
+          _::bitstring>> = left_binary
+
+        <<_::bitstring-size(right_offset_bits),
+          right_batch_item_binary::bitstring-size(right_batch_item_bits),
+          _::bitstring>> = right_binary
+
+        batch_item_output_template
+        |> dot4(
+          from_binary(left_batch_item_template, left_batch_item_binary),
+          left_batch_contract_axes,
+          from_binary(right_batch_item_template, right_batch_item_binary),
+          right_batch_contract_axes
+        )
+        |> to_binary()
```

Comment:
I would change bin_zip_reduce and dot4 (maybe call it `bin_dot4`) so it returns a binary instead of a tensor. This means you will no longer need to compute the `batch_item_output_template`.

Perhaps you should also change bin_reduce to return a binary too, for consistency.